nickel/src/lib.rs

```diff
@@ -0,0 +1,637 @@
+use malachite::base::{num::conversion::traits::RoundingFrom, rounding_modes::RoundingMode};
+
+use nickel_lang_core::{
+    error::{report::DiagnosticsWrapper, IntoDiagnostics, NullReporter},
+    eval::{cache::CacheImpl, Closure, Environment},
+    files::Files,
+    identifier::Ident,
+    program::Program,
+    serialize::{to_string, validate, ExportFormat},
+    term::{self, record::RecordData, RichTerm, Term},
+};
+
+pub use nickel_lang_core::error::report::ErrorFormat;
+
+/// Provides a destination for the output of `std.trace`.
+#[derive(Clone)]
+pub struct Trace {
+    // This is a little annoying, in that VirtualMachine already takes a generic
+    // trace and boxes it. Since we don't want a generic parameter on Context,
+    // it means we need to double-box it.
+    inner: Rc<RefCell<dyn Write>>,
+}
+
+impl Trace {
+    pub fn new<W: Write + 'static>(write: W) -> Self {
+        Trace {
+            inner: Rc::new(RefCell::new(write)),
+        }
+    }
+}
+
+impl Write for Trace {
+    fn write(&mut self, buf: &[u8]) -> std::io::Result<usize> {
+        self.inner.borrow_mut().write(buf)
+    }
+
+    fn flush(&mut self) -> std::io::Result<()> {
+        self.inner.borrow_mut().flush()
+    }
+}
+
+impl Default for Trace {
+    fn default() -> Trace {
+        Trace::new(std::io::sink())
+    }
+}
+
+/// The main entry point.
+#[derive(Clone, Default)]
+pub struct Context {
+    import_paths: Vec<OsString>,
+    trace: Trace,
+    name: Option<String>,
+}
+
+/// Controls the usage of ANSI escape codes in error messages.
+pub enum Color {
+    Ansi,
+    Off,
+}
+
+/// A Nickel evaluation error.
+pub struct Error {
+    error: Box<nickel_lang_core::error::Error>,
+    files: nickel_lang_core::files::Files,
+}
+
+impl std::fmt::Debug for Error {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        self.error.fmt(f)
+    }
+}
+
+impl Error {
+    pub fn format<W: Write>(
+        mut self,
+        write: &mut W,
+        format: ErrorFormat,
+        color: Color,
+    ) -> Result<(), Box<dyn std::error::Error>> {
+        let diagnostics = DiagnosticsWrapper::from(self.error.into_diagnostics(&mut self.files));
+        match format {
+            ErrorFormat::Text => {
+                let mut ansi;
+                let mut no_color;
+                let writer: &mut dyn WriteColor = match color {
+                    Color::Ansi => {
+                        ansi = Ansi::new(write);
+                        &mut ansi
+                    }
+                    Color::Off => {
+                        no_color = NoColor::new(write);
+                        &mut no_color
+                    }
+                };
+                diagnostics.diagnostics.iter().try_for_each(|d| {
+                    Ok(codespan_reporting::term::emit(
+                        writer,
+                        &codespan_reporting::term::Config::default(),
```

Comment:
this is probably optimized away, but maybe we could pull the creation of the config out of the loop